languages/rust/src/query.rs

```diff
@@ -0,0 +1,282 @@
+                    attribute,
+                    args,
+                } => self.handle_external_call(call_id, instance, attribute, args),
+                QueryEvent::ExternalOp {
+                    call_id,
+                    operator,
+                    args,
+                } => self.handle_external_op(call_id, operator, args),
+                QueryEvent::ExternalIsa {
+                    call_id,
+                    instance,
+                    class_tag,
+                } => self.handle_external_isa(call_id, instance, class_tag),
+                QueryEvent::ExternalUnify {
+                    call_id,
+                    left_instance_id,
+                    right_instance_id,
+                } => self.handle_external_unify(call_id, left_instance_id, right_instance_id),
+                QueryEvent::ExternalIsSubSpecializer {
+                    call_id,
+                    instance_id,
+                    left_class_tag,
+                    right_class_tag,
+                } => self.handle_external_is_subspecializer(
+                    call_id,
+                    instance_id,
+                    left_class_tag,
+                    right_class_tag,
+                ),
+                QueryEvent::Debug { message } => self.handle_debug(message),
+            };
+            if let Err(e) = result {
+                self.application_error(e);
+            }
+        }
+    }
+
+    fn check_messages(&mut self) {
+        while let Some(msg) = self.inner.next_message() {
+            match msg.kind {
+                polar_core::types::MessageKind::Print => tracing::trace!("{}", msg.msg),
+                polar_core::types::MessageKind::Warning => tracing::warn!("{}", msg.msg),
+            }
+        }
+    }
+
+    fn question_result(&mut self, call_id: u64, result: bool) {
+        self.inner.question_result(call_id, result);
+    }
+
+    fn call_result(&mut self, call_id: u64, result: Arc<dyn ToPolar>) -> anyhow::Result<()> {
+        let mut host = self.host.lock().unwrap();
+        let value = host.value_to_polar(result.as_ref());
+        Ok(self.inner.call_result(call_id, Some(value))?)
+    }
+
+    fn call_result_none(&mut self, call_id: u64) -> anyhow::Result<()> {
+        Ok(self.inner.call_result(call_id, None)?)
+    }
+
+    fn application_error(&mut self, error: anyhow::Error) {
+        self.inner.application_error(error.to_string())
+    }
+
+    fn handle_make_external(&mut self, instance_id: u64, constructor: Term) -> anyhow::Result<()> {
+        let mut host = self.host.lock().unwrap();
+        match constructor.value() {
+            Value::InstanceLiteral(InstanceLiteral { .. }) => todo!("instantiate from literal"),
+            Value::Call(Predicate { name, args }) => {
+                let _instance = host.make_instance(name, args.clone(), instance_id);
+            }
+            _ => panic!("not valid"),
+        }
+        Ok(())
+    }
+
+    fn register_call(
```

Comment:
Nit: argument order is different from other libs.

Reply:
It's a good nit. My logic was:

`call_id: instance.name(args)` is roughly how you would read this